app/uk/gov/hmrc/channelpreferences/controllers/PreferenceController.scala

```diff
@@ -76,4 +83,100 @@ class PreferenceController @Inject()(
+             | Case 1.4 - SAUTR does not exist in the  Auth token
+             |
+             |   Given I am a customer who has successfully enrolled in ITSA
+             |   When SAUTR does not exist in the Auth token
+             |   Then my itsaId will be added (i.e linked to) the entityId
+             */
+            entityResolverConnector
+              .update(resolvedEntity.copy(itsa = Some(passedBack_itsaId)))
+              .map { _ =>
+                Ok("itsaId successfully linked to entityId")
+              }
+          } else {
+            // authToken_saUtr.isDefined
+            if (resolvedEntity.saUtr.isDefined && (resolvedEntity.saUtr.get == authToken_saUtr.get)) {
+              /*
+               | Case 1.1 - SAUTR in Auth token is the same as SAUTR in entity resolver
+               |
+               |   Given I am a customer who has successfully enrolled in ITSA
+               |   When SAUTR in Auth token is the same as SAUTR in entity resolver
+               |   Then my itsaId will be added (i.e linked to) the entityId
+               */
+              entityResolverConnector
+                .update(resolvedEntity.copy(itsa = Some(passedBack_itsaId)))
+                .map { _ =>
+                  Ok("itsaId successfully linked to entityId")
+                }
+            } else {
+              /*
+               | Case 1.2 - SAUTR in Auth token is different from  SAUTR in entity resolver
+               |
+               |   Given I am a customer who has successfully enrolled in ITSA
+               |   When SAUTR in Auth token is different from  SAUTR in entity resolver
+               |   Then my itsaId will not be added (i.e linked to) any entityId
+               |   And an error will be generated
+               */
+              Future.successful(Unauthorized("SAUTR in Auth token is different from SAUTR in entity resolver"))
+
+              // TODO Isn't the following acceptance criteria already included in Case 1.2 ???
+              // Case 1.3 - SAUTR in Auth token is linked to a different entityId  in entity resolver
+              //
+              //   Given I am a customer who has successfully enrolled in ITSA
+              //   When SAUTR in Auth token is linked to a different entityId  in entity resolver
+              //   Then my itsaId will not be added (i.e linked to) any entityId
+              //   And an error will be generated
+            }
+          }
+        }
+      }
+      .recoverWith {
+        case _: Throwable =>
+          /*
+           | Case 3.1 - entityId passed back by ITSA does not exist in entity resolver
+           |
+           |   Given I am a customer who has successfully enrolled in ITSA
+           |   When the entityId passed back does not exist in entity resolver
+           |   Then  my itsaId will not be added to any entityid
+           |   And an error will be generated
+           */
+          Future.successful(NotFound("Entity ID not found"))
```

Comment:
Not found on any throwable? Is this really the case?

Reply:
I'm not sure but maybe you can restrict it to `AuthorisationException` only ?

Reply:
After drilling down into the HRMC libraries, I guess, the best I can write probably is the following pattern match:

```scala 

import uk.gov.hmrc.http.UpstreamErrorResponse

// Interact with the remote entity-resolver service
entityResolverConnector
      .resolveBy(passedBackEntityId)
      // .flatMap { entity => ... }
      .recoverWith {
         case ex: UpstreamErrorResponse if (ex.statusCode == NOT_FOUND) =>
            reply(NOT_FOUND, "Invalid entity id or entity id has expired")

         case ex: Throwable =>
            reply(INTERNAL_SERVER_ERROR, ex.getMessage)
       }
```

What would YOU write in similar circumstances, please?